## **Notebook 1: Data Cleaning & Wrangling**

Our main objective in this notebook is to transform raw Airbnb Berlin dataset into a clean, reliable, and analysis-ready dataset.

For this purpose, we:
-  Load the original `listings.csv` file
-  Select key columns/features for business-focused analysis
-  Explore data quality and identify missing values
-  Clean, standardize, and transform important features
-  Handle missing and inconsistent data
-  Engineer new features for deeper insights
-  Export the cleaned dataset as `listings_cleaned.csv` for further analysis

### **1- Setup and Load Data**
---

In [1]:
# Import necessary libraries
import pandas as pd 
import os

In [2]:
# For better readability and show all columns
pd.set_option('display.max_columns', None)

In [3]:
# Load the listing dataset
path = '../data/listings.csv'

try:
    df = pd.read_csv(path)
    print("Data loaded successfully!")
except FileNotFoundError: 
    print("File not found! Check the path!")

Data loaded successfully!


### **2- Inspect the Dataset**
---

In [4]:
# Quick preview of data
df.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,host_url,host_name,host_since,host_location,host_about,host_response_time,host_response_rate,host_acceptance_rate,host_is_superhost,host_thumbnail_url,host_picture_url,host_neighbourhood,host_listings_count,host_total_listings_count,host_verifications,host_has_profile_pic,host_identity_verified,neighbourhood,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,property_type,room_type,accommodates,bathrooms,bathrooms_text,bedrooms,beds,amenities,price,minimum_nights,maximum_nights,minimum_minimum_nights,maximum_minimum_nights,minimum_maximum_nights,maximum_maximum_nights,minimum_nights_avg_ntm,maximum_nights_avg_ntm,calendar_updated,has_availability,availability_30,availability_60,availability_90,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,availability_eoy,number_of_reviews_ly,estimated_occupancy_l365d,estimated_revenue_l365d,first_review,last_review,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,3176,https://www.airbnb.com/rooms/3176,20250315152850,2025-03-16,city scrape,Fabulous Flat in great Location,This beautiful first floor apartment is situa...,The neighbourhood is famous for its variety of...,https://a0.muscache.com/pictures/airflow/Hosti...,3718,https://www.airbnb.com/users/show/3718,Britta,2008-10-19,"Coledale, Australia",We love to travel ourselves a lot and prefer t...,within a day,67%,40%,f,https://a0.muscache.com/im/users/3718/profile_...,https://a0.muscache.com/im/users/3718/profile_...,Prenzlauer Berg,1.0,1.0,"['email', 'phone']",t,t,"Berlin, Germany",Prenzlauer Berg Südwest,Pankow,52.53471,13.41810,Entire rental unit,Entire home/apt,2,1.0,1 bath,1.0,2.0,"[""Bathtub"", ""Dishes and silverware"", ""Bed line...",$105.00,63,730,1.0,63.0,730.0,730.0,62.7,730.0,NaN,t,0,0,11,286,2025-03-16,149,1,1,212,0,126,13230.0,2009-06-20,2025-03-02,4.63,4.67,4.52,4.65,4.70,4.92,4.61,First name and Last name: Nicolas Krotz <br/> ...,f,1,1,0,0,0.78
1,9991,https://www.airbnb.com/rooms/9991,20250315152850,2025-03-16,city scrape,Geourgeous flat - outstanding views,4 bedroom with very large windows and outstand...,Prenzlauer Berg is an amazing neighbourhood wh...,https://a0.muscache.com/pictures/42799131/59c8...,33852,https://www.airbnb.com/users/show/33852,Philipp,2009-08-25,"Berlin, Germany",Born in Munich - lives in and loves Berlin. Wo...,a few days or more,0%,0%,f,https://a0.muscache.com/im/users/33852/profile...,https://a0.muscache.com/im/users/33852/profile...,Prenzlauer Berg,1.0,1.0,"['email', 'phone']",t,t,"Berlin, Germany",Prenzlauer Berg Südwest,Pankow,52.53269,13.41805,Entire rental unit,Entire home/apt,7,2.5,2.5 baths,4.0,4.0,"[""Bathtub"", ""Dryer"", ""Dishes and silverware"", ...",$135.00,6,14,6.0,6.0,14.0,14.0,6.0,14.0,NaN,t,0,0,0,109,2025-03-16,7,0,0,35,0,0,0.0,2015-08-09,2020-01-04,5.00,5.00,5.00,5.00,5.00,4.86,4.86,03/Z/RA/003410-18,f,1,1,0,0,0.06
2,14325,https://www.airbnb.com/rooms/14325,20250315152850,2025-03-16,city scrape,Studio Apartment in Prenzlauer Berg,The apartment is located on the upper second f...,NaN,https://a0.muscache.com/pictures/508703/24988a...,55531,https://www.airbnb.com/users/show/55531,Chris + Oliver,2009-11-18,"Berlin, Germany","From Berlin, Germany. We love to travel!",within a few hours,100%,38%,t,https://a0.muscache.com/im/users/55531/profile...,https://a0.muscache.com/im/users/55531/profile...,Prenzlauer Berg,4.0,5.0,"['email', 'phone', 'work_email']",t,t,NaN,Prenzlauer Berg Nordwest,Pankow,52.54813,13.40366,Entire rental unit,Entire home/apt,1,1.0,1 bath,0.0,1.0,"[""Heating"", ""Host greets 

In [5]:
# Dataset shape and column names
df.shape, df.columns

((13945, 79),
 Index(['id', 'listing_url', 'scrape_id', 'last_scraped', 'source', 'name',
        'description', 'neighborhood_overview', 'picture_url', 'host_id',
        'host_url', 'host_name', 'host_since', 'host_location', 'host_about',
        'host_response_time', 'host_response_rate', 'host_acceptance_rate',
        'host_is_superhost', 'host_thumbnail_url', 'host_picture_url',
        'host_neighbourhood', 'host_listings_count',
        'host_total_listings_count', 'host_verifications',
        'host_has_profile_pic', 'host_identity_verified', 'neighbourhood',
        'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'latitude',
        'longitude', 'property_type', 'room_type', 'accommodates', 'bathrooms',
        'bathrooms_text', 'bedrooms', 'beds', 'amenities', 'price',
        'minimum_nights', 'maximum_nights', 'minimum_minimum_nights',
        'maximum_minimum_nights', 'minimum_maximum_nights',
        'maximum_maximum_nights', 'minimum_nights_avg_ntm',
        

In [6]:
# Data types and non-null value counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13945 entries, 0 to 13944
Data columns (total 79 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            13945 non-null  int64  
 1   listing_url                                   13945 non-null  object 
 2   scrape_id                                     13945 non-null  int64  
 3   last_scraped                                  13945 non-null  object 
 4   source                                        13945 non-null  object 
 5   name                                          13945 non-null  object 
 6   description                                   13350 non-null  object 
 7   neighborhood_overview                         6612 non-null   object 
 8   picture_url                                   13945 non-null  object 
 9   host_id                                       13945 non-null 

In [7]:
# Check for duplicates in the dataset
df.duplicated().sum()

0

### **3- Select Columns Relevant to Our Business Questions**
---

In [8]:
# List of columns we want to keep
columns_to_keep = [
    'price',                        # Price analysis
    'neighbourhood_cleansed',       # Neighbourhood-level trends
    'neighbourhood_group_cleansed', # Neighbourhood-level trends
    'latitude', 'longitude',        # Geospatial mapping
    'room_type',                    # Room type analysis
    'availability_30',              # Number of available nights in the next 30 days
    'host_is_superhost',            # Superhost analysis
    'host_since',                   # Host experience
    'number_of_reviews',            # Demand proxy
    'review_scores_rating',         # Quality analysis
    'amenities'                     # Amenities analysis
]

In [9]:
# Check if all the columns we want to keep are in the dataset
missing = set(columns_to_keep) - set(df.columns)
if missing:
    raise KeyError(f"Expected columns missing from DataFrame: {missing}")

In [10]:
# Subset the dataset
df = df[columns_to_keep].copy()

print(f"\nAfter subsetting, here is the new dataset shape: {df.shape}")


After subsetting, here is the new dataset shape: (13945, 12)


In [11]:
df.head(3)

,price,neighbourhood_cleansed,neighbourhood_group_cleansed,latitude,longitude,room_type,availability_30,host_is_superhost,host_since,number_of_reviews,review_scores_rating,amenities
0,$105.00,Prenzlauer Berg Südwest,Pankow,52.53471,13.41810,Entire home/apt,0,f,2008-10-19,149,4.63,"[""Bathtub"", ""Dishes and silverware"", ""Bed line..."
1,$135.00,Prenzlauer Berg Südwest,Pankow,52.53269,13.41805,Entire home/apt,0,f,2009-08-25,7,5.00,"[""Bathtub"", ""Dryer"", ""Dishes and silverware"", ..."
2,$75.00,Prenzlauer Berg Nordwest,Pankow,52.54813,13.40366,Entire home/apt,0,t,2009-11-18,26,4.68,"[""Heating"", ""Host greets you"", ""Smoke alarm"", ..."


### **4- Data Completeness and Missing Value Overview**
---

In [12]:
# Check the quick info again for the new dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13945 entries, 0 to 13944
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   price                         8898 non-null   object 
 1   neighbourhood_cleansed        13945 non-null  object 
 2   neighbourhood_group_cleansed  13945 non-null  object 
 3   latitude                      13945 non-null  float64
 4   longitude                     13945 non-null  float64
 5   room_type                     13945 non-null  object 
 6   availability_30               13945 non-null  int64  
 7   host_is_superhost             13803 non-null  object 
 8   host_since                    13917 non-null  object 
 9   number_of_reviews             13945 non-null  int64  
 10  review_scores_rating          10631 non-null  float64
 11  amenities                     13945 non-null  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 1.3+ MB


In [13]:
# Check the missing values in the dataset
missing = df.isnull().sum().sort_values(ascending=False)
missing[missing > 0]

price                   5047
review_scores_rating    3314
host_is_superhost        142
host_since                28
dtype: int64

In [14]:
# Check the completeness percentage of the dataset
total_listings = df.shape[0]
completeness = (df.count() / total_listings * 100).round(1)
completeness.sort_values(ascending=False)

neighbourhood_cleansed          100.0
neighbourhood_group_cleansed    100.0
latitude                        100.0
longitude                       100.0
room_type                       100.0
availability_30                 100.0
number_of_reviews               100.0
amenities                       100.0
host_since                       99.8
host_is_superhost                99.0
review_scores_rating             76.2
price                            63.8
dtype: float64

A review of missing-value counts indicates:

- **Host Since** (`host_since`): ~99.8% complete (13,917/13,945). For data lacking a host_since date, we will assign an experience of 0 years when calculating `host_experience_years`.

- **Host Superhost Status** (`host_is_superhost`): ~99.0% complete (13,803/13,945). Missing entries will be imputed as `False`, assuming the absence of this flag indicates a regular (non-superhost) host.

- **Review Scores Rating** (`review_scores_rating`): ~76.2% complete (10,631/13,945). Missing review scores will be filled using the median value, preserving the central tendency and avoiding bias.

- **Nightly Price** (`price`): ~63.8% complete (8,898/13,945). As price is essential for our all downstream analyses, listings without a valid price will be excluded from further analysis.

- **Other Selected Columns**: All other columns chosen for analysis are 100% complete, and therefore they do not require further imputation.

### **5- Data Wrangling and Preparation**
---

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13945 entries, 0 to 13944
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   price                         8898 non-null   object 
 1   neighbourhood_cleansed        13945 non-null  object 
 2   neighbourhood_group_cleansed  13945 non-null  object 
 3   latitude                      13945 non-null  float64
 4   longitude                     13945 non-null  float64
 5   room_type                     13945 non-null  object 
 6   availability_30               13945 non-null  int64  
 7   host_is_superhost             13803 non-null  object 
 8   host_since                    13917 non-null  object 
 9   number_of_reviews             13945 non-null  int64  
 10  review_scores_rating          10631 non-null  float64
 11  amenities                     13945 non-null  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 1.3+ MB


#### 5.1- Format `price` column and handling missing values

In [16]:
# Convert 'price' column to numeric format, removing non-numeric characters in it
df['price'] = df['price'].str.replace('$', '', regex=False).str.replace(',', '', regex=False).astype(float)

# Drop rows with missing 'price' values
df = df.dropna(subset=['price'])

# Check price stats
print(df['price'].describe())

count     8898.000000
mean       128.763880
std        316.080704
min          9.000000
25%         65.000000
50%         94.000000
75%        138.000000
max      10006.000000
Name: price, dtype: float64


#### 5.2- Convert `host_since` column to datetime

In [17]:
#  Convert `host_since` to datetime, so missing dates are NaT
df['host_since'] = pd.to_datetime(df['host_since'], errors='coerce')
print(f"`host_since` dtype after conversion: {df['host_since'].dtype}")

`host_since` dtype after conversion: datetime64[ns]


#### 5.3- Impute missing `host_is_superhost` as False (not superhost)

In [18]:
# Fill missing 'host_is_superhost' column with 'f'
df['host_is_superhost'] = df['host_is_superhost'].fillna('f').astype('category')
df['host_is_superhost'].value_counts()

host_is_superhost
f    6065
t    2833
Name: count, dtype: int64

#### 5.4- Impute missing `review_scores_rating` with median

In [19]:
# Fill missing 'review_score_rating' column with the median value to maintain the distribution's central tendency
df['review_scores_rating'] = df['review_scores_rating'].fillna(df['review_scores_rating'].median())

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8898 entries, 0 to 13944
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   price                         8898 non-null   float64       
 1   neighbourhood_cleansed        8898 non-null   object        
 2   neighbourhood_group_cleansed  8898 non-null   object        
 3   latitude                      8898 non-null   float64       
 4   longitude                     8898 non-null   float64       
 5   room_type                     8898 non-null   object        
 6   availability_30               8898 non-null   int64         
 7   host_is_superhost             8898 non-null   category      
 8   host_since                    8879 non-null   datetime64[ns]
 9   number_of_reviews             8898 non-null   int64         
 10  review_scores_rating          8898 non-null   float64       
 11  amenities                     8898

#### 5.5- Convert `room_type`, `neighbourhood_cleansed`, and `neighbourhood_group_cleansed` columns to pandas category dtype

In [21]:
# Convert categorical columns to category dtype
for col in ['room_type', 'neighbourhood_cleansed', 'neighbourhood_group_cleansed']:
    df[col] = df[col].astype('category')

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8898 entries, 0 to 13944
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   price                         8898 non-null   float64       
 1   neighbourhood_cleansed        8898 non-null   category      
 2   neighbourhood_group_cleansed  8898 non-null   category      
 3   latitude                      8898 non-null   float64       
 4   longitude                     8898 non-null   float64       
 5   room_type                     8898 non-null   category      
 6   availability_30               8898 non-null   int64         
 7   host_is_superhost             8898 non-null   category      
 8   host_since                    8879 non-null   datetime64[ns]
 9   number_of_reviews             8898 non-null   int64         
 10  review_scores_rating          8898 non-null   float64       
 11  amenities                     8898

### **6- Feature Engineering**
---

#### 6.1- Compute host experience in years

In [23]:
# Compute experience in years (NaT values will yield NaN)
df['host_experience_years'] = (pd.Timestamp.today() - df['host_since']).dt.days / 365

# Fill missing experience with 0
df['host_experience_years'] = df['host_experience_years'].fillna(0)

# Show `host_experience_year` stats after cleaning
df['host_experience_years'].describe()

count    8898.000000
mean        7.685339
std         4.069817
min         0.000000
25%         3.836301
50%         8.126027
75%        11.043836
max        16.923288
Name: host_experience_years, dtype: float64

#### 6-2. Parse amenities into a list and count

In [24]:
# Parse amenities list from string, removing braces and quotes
df['amenities_list'] = df['amenities'].str.strip('[]').str.replace('"', '').str.split(',')

# Standardize amenity names (lowercase, strip spaces)
df['amenities_list'] = df['amenities_list'].apply(lambda x: [a.strip().lower() for a in x])

# Count number of amenities for each listing
df['amenities_count'] = df['amenities_list'].apply(len)

In [ ]:
df.isnull().sum()

price                            0
neighbourhood_cleansed           0
neighbourhood_group_cleansed     0
latitude                         0
longitude                        0
room_type                        0
availability_30                  0
host_is_superhost                0
host_since                      19
number_of_reviews                0
review_scores_rating             0
amenities                        0
host_experience_years            0
amenities_list                   0
amenities_count                  0
dtype: int64

**Note on `host_since` missing values:** 

All columns, except `host_since`, are complete with no missing values.  
This is not a problem for our analysis, as we rely on the derived feature `host_experience_years` (years since host joined) for EDA.  
For listings with a missing `host_since` date, `host_experience_years` is set to 0, ensuring no gaps in downstream analyses.  
We retain the original `host_since` column for transparency and to enable any future analyses on host tenure.

### **7- Export the cleaned dataset**
---

In [26]:
df.to_csv('../data/listings_cleaned.csv', index=False)
print(f"Exported cleaned dataset: 'data/listings_cleaned.csv' (Shape: {df.shape})")

Exported cleaned dataset: 'data/listings_cleaned.csv' (Shape: (8898, 15))


✅ The cleaned Airbnb dataset has been exported as `listings_cleaned.csv` and is now ready for exploratory data analysis.